lib/ControlSystemsBase/src/types/conversion.jl

```diff
@@ -298,18 +299,40 @@ function siso_ss_to_zpk(sys, i, j)
     return z, eigvals(A), k
 end
 
+function charpoly(A::AbstractMatrix{T}) where {T}
+    N = size(A, 1)
+    poly_factors::Vector{T} = vec(ones(T, N+1, 1))
+    if N <= 0
+        return poly_factors
+    end
 
-# TODO: Could perhaps be made more accurate. See: An accurate and efficient
-# algorithm for the computation of the # characteristic polynomial of a general square matrix.
-function charpoly(A::AbstractMatrix{T}) where T
-    Λ::Vector{T} = eigvalsnosort(A)
-    return prod(roots2poly_factors(Λ))::Polynomial{T,:x} # Compute the polynomial factors directly?
-end
-function charpoly(A::AbstractMatrix{<:Real})
-    Λ = eigvalsnosort(A)
-    return prod(roots2real_poly_factors(Λ))
+    t::Matrix{T} = zeros(N, N) # Preallocation
```

Comment:
```suggestion
    t = zeros(T, N, N) # Preallocation
```
the element type of a zero matrix is specified like this